### ISSUE TITLE
Loosing Gromacs atom name details when using an OpenMM topology

### ISSUE THREAD

**Issue by hannahbaumann** [openforcefield/openff-interchange#1031]
When I create an interchange using `from_openmm` and I provide an OpenMM topology for the `topology=` option, and then write out the Gromacs .top and .gro file, the atom names of protein atoms are, e.g. C and not CA. When I instead provide an OpenFF topology in the `from_openmm` call, the protein atom names have the same names as in the .pdb file used, e.g. CA. Is this information lost in the OpenMM topology or would it be possible to also transfer the atom names through that one?

**Comment by mattwthompson** [comment]
This is legit and just an oversight; most of my testing has ensured that energy evaluations are reproduced in these code paths and atom names aren't needed for that.

The short of it, I think is - 

* The toolkit has `Topology.from_openmm` / `Topology.from_pdb` that Jeff and others have invested a lot of effort into and probably does what users expect w/r/t atom names, so if you used the toolkit to load a PDB file the atom names are probably accurately defined
* Interchange has a thinner function that does the same thing when an OpenMM topology is passed to `from_openmm`, but does not route through the toolkit's solution and probably failed to process atom names from OpenMM for no reason other than oversight

I think I'm able to isolate what you're running into:

```python
"""
$ INTERCHANGE_EXPERIMENTAL=1 python issue_1031.py
"""

import openmm.app
from openff.toolkit import Topology
from openff.interchange import Interchange


def get_atom_names(topology: Topology) -> set[str]:
    return {atom.name for atom in topology.atoms}


pdb_file = openmm.app.PDBFile("protein.pdb")

interchange = Interchange.from_openmm(
    system=openmm.app.ForceField(
        "amber99sb.xml",
        "tip3p.xml",
    ).createSystem(
        pdb_file.topology,
        nonbondedMethod=openmm.app.PME,
    ),
    topology=pdb_file.topology,
)

# These are more likely to accurately represent the intended data in the PDB;
# JW/JM and friends put a lot of work into this
toolkit_topology = Topology.from_pdb("protein.pdb")

print(get_atom_names(toolkit_topology))
# {'HD13', 'NZ', 'CG1', 'HD21', 'HG21', 'HB1', 'HG2', 'HG23', 'HD12', 'HE2', 'OH', 'HA', 'H1', 'HA3', 'H3', 'HD11',
# 'ND2', 'CZ', 'HZ', 'HD23', 'HH12', 'OG', 'HH21', 'CH2', 'C', 'OE1', 'H2', 'O', 'CD1', 'CE3', 'HB3', 'HE1', 'HB2',
# 'HE3', 'NH1', 'CG2', 'CA', 'OE2', 'HZ2', 'SD', 'HH11', 'SG', 'HG1', 'H', 'CE2', 'HZ3', 'HG13', 'HD22', 'HH', 'HH22',
# 'CZ2', 'HH2', 'CG', 'NH2', 'CE', 'N', 'CB', 'HD3', 'HE22', 'CD', 'HE', 'OD1', 'HD1', 'ND1', 'HA2', 'NE', 'NE1',
# 'CH3', 'HG11', 'HG22', 'HD2', 'HB', 'NE2', 'HG12', 'OG1', 'HE21', 'CE1', 'CD2', 'HG3', 'CZ3', 'HG', 'HZ1', 'OD2'}

# These match what OpenMM reports, so probably correct?
assert get_atom_names(toolkit_topology) == {atom.name for atom in pdb_file.topology.atoms()}

# Interchange's representation doesn't ingest these atom names
print(get_atom_names(interchange.topology))
# {''}
```

1. Does this accurately capture your expectations, and how Interchange loses the atom name information?
2. Would simply setting atom names from the `openmm_topology.atoms()` iterator suffice here, or are there other complications? (Stuff like [this](https://docs.openforcefield.org/projects/toolkit/en/stable/users/molecule_conversion.html) scares me, but I think atom names are simpler than residue/chain/insertion code spookiness.)
3. Are there other related bits of information or metadata that are missed that you'd expect not to be missed?

**Comment by hannahbaumann** [comment]
Thank you @mattwthompson for looking into this!
1. Yes, that is exactly what I observed!
2. Yes, I think that sounds like a good idea!
3. I can't think of other things!

**Comment by mattwthompson** [comment]
Perfect, thanks. I'll work on the patch soon - hopefully it shakes out to be as simple as it seems to be right now

**Comment by mattwthompson** [comment]
The fix is simple (one short line of code in the right place) but not so simple to rollout because it's actually in the toolkit (https://github.com/openforcefield/openff-toolkit/pull/1927). I have a fix in place (https://github.com/openforcefield/openff-toolkit/pull/1928) which should fix this - I'll run a few tests before making the release, but I don't expect any further hiccups.

As a matter of safety I don't make releases on Fridays, but this is looking good to launch early next week. Anticipate needing to upgrade to toolkit 0.16.4 and not needing to upgrade Interchange.

**Comment by mattwthompson** [comment]
I think this is fixed with `openff-toolkit =0.16.4`


```python
In [2]: print(get_atom_names(toolkit_topology))
{'NE', 'HD22', 'HA3', 'HD3', 'HZ1', 'CE3', 'HH21', 'HG12', 'CE', 'OE2', 'HG21', 'OD1', 'HA', 'HZ3', 'HZ', 'CB', 'NH2', 'ND2', 'HD12', 'CG', 'CD1', 'OH', 'CZ', 'HB3', 'HG23', 'HG', 'HE2', 'HD21', 'OD2', 'HD11', 'H2', 'O', 'OE1', 'HH12', 'HE3', 'HG13', 'HD13', 'HD1', 'HH', 'NE1', 'H3', 'HE1', 'SD', 'H', 'HE22', 'HG2', 'CE1', 'OG1', 'HH2', 'CH3', 'NE2', 'NZ', 'CZ2', 'CZ3', 'ND1', 'CA', 'N', 'CG1', 'SG', 'HA2', 'HB2', 'CD', 'CD2', 'HB', 'HG1', 'HG11', 'HB1', 'C', 'HH22', 'CG2', 'CH2', 'H1', 'HD2', 'HG22', 'HE', 'HZ2', 'HE21', 'HD23', 'HH11', 'HG3', 'NH1', 'CE2', 'OG'}

In [3]: print(get_atom_names(interchange.topology))
   ...:
{'NE', 'HD22', 'HA3', 'HD3', 'HZ1', 'CE3', 'HH21', 'HG12', 'CE', 'OE2', 'HG21', 'OD1', 'HA', 'HZ3', 'HZ', 'CB', 'NH2', 'ND2', 'HD12', 'CG', 'CD1', 'OH', 'CZ', 'HB3', 'HG23', 'HG', 'HE2', 'HD21', 'OD2', 'HD11', 'H2', 'O', 'OE1', 'HH12', 'HE3', 'HG13', 'HD13', 'HD1', 'HH', 'NE1', 'H3', 'HE1', 'SD', 'H', 'HE22', 'HG2', 'CE1', 'OG1', 'HH2', 'CH3', 'NE2', 'NZ', 'CZ2', 'CZ3', 'ND1', 'CA', 'N', 'CG1', 'SG', 'HA2', 'HB2', 'CD', 'CD2', 'HB', 'HG1', 'HG11', 'HB1', 'C', 'HH22', 'CG2', 'CH2', 'H1', 'HD2', 'HG22', 'HE', 'HZ2', 'HE21', 'HD23', 'HH11', 'HG3', 'NH1', 'CE2', 'OG'}
